src/util/RtreeBuild.cpp

```diff
@@ -0,0 +1,1133 @@
+
+      if (((ySize + 1) % currentS == 0 && (ySize + 1) / currentS >= 1 && (ySize + 1) / currentS < M)
+          || (ySize % currentS == 0 && ySize / currentS >= 1 && ySize / currentS < M)) {
+        // index i * S - 1 or i * S
+        r1Small->push_back(element);
+      }
+
+      if (ySize == 0) {
+        minD1 = element;
+        maxD1 = element;
+      }
+      if (element.orderY > maxD1.orderY) {
+        maxD1 = element;
+      }
+
+      ySize++;
+    }
+    centerOrdering(*RectanglesD0WithOrder, 0);
+  } else {
+    for (RTreeValueWithOrderIndex element : sorterRectsD1.sortedView()) {
+      element.orderY = ySize;
+      Rtree::SaveEntryWithOrderIndex(element, r1File);
+      sorterRectsD0.push(element);
+
+      if (((ySize + 1) % currentS == 0 && (ySize + 1) / currentS >= 1 && (ySize + 1) / currentS < M)
+          || (ySize % currentS == 0 && ySize / currentS >= 1 && ySize / currentS < M)) {
+        // index i * S - 1 or i * S
+        r1Small->push_back(element);
+      }
+
+      if (ySize == 0) {
+        minD1 = element;
+        maxD1 = element;
+      }
+      if (element.orderY > maxD1.orderY) {
+        maxD1 = element;
+      }
+
+      ySize++;
+    }
+  }
+
+  r1File.close();
+  sorterRectsD1.clear();
+
+  // replace the placeholder
+  (*r1Small)[0] = minD1;
+  (*r1Small)[1] = maxD1;
+
+  uint64_t currentX = 0;
+  std::ofstream r0File = std::ofstream(onDiskBase + ".boundingbox.d0.tmp", std::ios::binary);
+  std::shared_ptr<multiBoxWithOrderIndex> r0Small = std::make_shared<multiBoxWithOrderIndex>();
+  // placeholder
+  r0Small->push_back(RTreeValueWithOrderIndex());
+  r0Small->push_back(RTreeValueWithOrderIndex());
+  RTreeValueWithOrderIndex minD0;
+  RTreeValueWithOrderIndex maxD0;
+
+  if (workInRam) {
+    for (RTreeValueWithOrderIndex element : *RectanglesD0WithOrder) {
+      if (((currentX + 1) % currentS == 0 && (currentX + 1) / currentS >= 1 && (currentX + 1) / currentS < M)
+          || (currentX % currentS == 0 && currentX / currentS >= 1 && currentX / currentS < M)) {
+        // index i * S - 1 or i * S
+        r0Small->push_back(element);
+      }
+
+      if (currentX == 0) {
+        minD0 = element;
+        maxD0 = element;
+      }
+      if (element.orderX > maxD0.orderX) {
+        maxD0 = element;
+      }
+
+      currentX++;
+    }
+  } else {
+    for (RTreeValueWithOrderIndex element : sorterRectsD0.sortedView()) {
+      Rtree::SaveEntryWithOrderIndex(element, r0File);
+
+      if (((currentX + 1) % currentS == 0 && (currentX + 1) / currentS >= 1 && (currentX + 1) / currentS < M)
+          || (currentX % currentS == 0 && currentX / currentS >= 1 && currentX / currentS < M)) {
+        // index i * S - 1 or i * S
+        r0Small->push_back(element);
+      }
+
+      if (currentX == 0) {
+        minD0 = element;
+        maxD0 = element;
+      }
+      if (element.orderX > maxD0.orderX) {
+        maxD0 = element;
+      }
+
+      currentX++;
+    }
+  }
```

Comment:
Code duplication again (for the RAM as well as for the Dimension case). First think about it yourself and if you are stuck, concact me.